vm/chef/cookbooks/singlefs/files/singlefs-validate

```diff
@@ -0,0 +1,250 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+# Test that the filesystem, and optionally NFS/Samba works correctly.
+
+if (( "$(id -u)" > 0 )); then
+  echo "Please run this script as root."
+  exit 1
+fi
+
+# Don't echo commands to the command line / serial console.
+set +x
+
+# Only validate installation once.
+readonly COMPLETE_FILE=/opt/c2d/validation.complete
+if [[ -f "${COMPLETE_FILE}" ]]; then
+  echo "Software validation already completed."
+  exit 0
+fi
+touch "${COMPLETE_FILE}"
+
+. /opt/c2d/get_metadata.sh
+. /opt/c2d/instance_utils.sh
+
+# Redirect all output to the log file.
+readonly LOG_FILE=/opt/c2d/validation.log
+exec > >(tee -a "${LOG_FILE}" ) 2>&1
+
+set -o nounset
+set -o errexit
+set -o pipefail
+
+FAILED=0
+
+function passed {
+  echo "Passed."
+}
+
+function failed {
+  echo "Failed."
+  FAILED=1
+}
+
+function validate {
+  $(eval $1)
+  if [ $? -eq 0 ] ; then
+    passed
+  else
+    failed
+  fi
+}
+
+function test_file {
+  echo -n "Test that test file doesn't already exist ... "
+  validate "echo Hello > '$1'"
+
+  echo -n "Test that writing a file works ... "
+  validate "echo Hello > '$1'"
+
+  echo -n "Test that reading a file works ... "
+  validate "grep Hello '$1' > /dev/null"
+
+  echo -n "Test that deleting a file works ... "
+  validate "rm '$1' > /dev/null"
+}
+
+function test_nfs {
+  echo -n "Test that NFS server is responding ... "
+  validate "rpcinfo -t 127.0.0.1 100003 3 > /dev/null"
+
+  echo -n "Test that NFS is exported ... "
+  validate "exportfs > /dev/null"
+
+  echo -n "Test that NFS can be mounted ... "
+  validate "mount '$1:/${STORAGE_POOL}' /mnt"
+
+  # Only test file creation/deletion on primary (replicas are read only)
+  if is_primary; then
+    test_file /mnt/nfs_test.txt
+  fi
+
+  echo -n "Test that NFS can be unmounted ... "
+  validate "umount /mnt"
+}
+
+function test_grafana_dashboards {
+  local url="http://admin:${ADMIN_PASSWORD}@127.0.0.1:3000/api/search?query=&starred=false"
+  local dashboards=$(curl -s "${url}")
+  # Expect "Storage" and "System" dashboards.
+  if [[ "${dashboards}" =~ "Storage" ]] && [[ "${dashboards}" =~ "System" ]]; then
+    passed
+  else
+    failed
+  fi
+}
+
+echo -e "Wait 90s for NFS server to start and exit grace period.\n"
+sleep 90
+
+echo -e "Test installation and configuration.\n"
+
+readonly HOST_IP=$(hostname -i)
+readonly ADMIN_PASSWORD="$(get_metadata instance/attributes/ADMIN_PASSWORD)"
+readonly ENABLE_NFS="$(get_metadata instance/attributes/ENABLE_NFS)"
+readonly ENABLE_SMB="$(get_metadata instance/attributes/ENABLE_SMB)"
+readonly FILE_SYSTEM="$(get_metadata instance/attributes/FILE_SYSTEM)"
+readonly STORAGE_POOL="$(get_metadata instance/attributes/STORAGE_POOL_NAME)"
+readonly DISABLE_MONITORING="$(get_metadata instance/attributes/DISABLE_MONITORING || true)"
```

Comment:
**(Postpone action):** `DISABLE_MONITORING` attribute does not exist. Can be removed?